### ISSUE TITLE
Softmax Autograd Support for Discrete SAC to Match PyTorch

### ISSUE THREAD

**Issue by zhuruicheng** [rindow/rindow-neuralnetworks#14]
I am developing a discrete Soft Actor-Critic (SAC) reinforcement learning algorithm using Rindow NeuralNetworks. Having successfully implemented the continuous action version, I plan to start developing the discrete version of sac. However, the Softmax layer for discrete action probabilities fails to match PyTorch’s torch.softmax output ([[0.090031, 0.244728, 0.665241], [0.090031, 0.244728, 0.665241]] for input [[1.0, 2.0, 3.0], [0.5, 1.5, 2.5]]), and instead produces a uniform distribution [[0.333333, 0.333333, 0.333333, 0.333333]], because $K->maximum($x, 0.0) incorrectly returns the input and lacks automatic differentiation for gradient computation, which is crucial for SAC policy optimization. I'm requesting to add Softmax automatic differentiation support to Gradient to ensure the forward pass matches PyTorch and supports gradients. I'd be happy to help with testing or provide more details!

refer to:
`import torch
 
logits = torch.tensor([
    [1.0, 2.0, 3.0],
    [0.5, 1.5, 2.5]
], dtype=torch.float32)
 
probs = torch.softmax(logits, dim=1)  
 
print("PyTorch softmax:")
print(probs.round(decimals=6).tolist())  ` 

My php code is wrong:
`
        $mo = new MatrixOperator();
        $nn = new NeuralNetworks($mo);
        $K = $nn->backend();

       
        $x = $K->array([[1.0, 2.0, 3.0], [0.5, 1.5, 2.5]], NDArray::float32);
        $max = $K->maximum($x, 0.0); // Incorrect: returns $x
        $sub = $K->sub($x, $max); // Results in zeros: [[0, 0, 0], [0, 0, 0]]
        $expX = $K->exp($sub); // [[1, 1, 1], [1, 1, 1]]
        $sumExpX = $K->sum($expX, $axis=1, $keepdims=true); // [[3], [3]]
        $sumExpX_expanded = $K->concat(array_fill(0, 3, $sumExpX), $axis=1);
        $output = $K->div($expX, $sumExpX_expanded); // [[0.333333, 0.333333, 0.333333], [0.333333, 0.333333, 0.333333]]
        echo json_encode($output->toArray()).PHP_EOL;
`

**Comment by yuichiis** [comment]
The calculation dimension axis for softmax is always -1, which means it's the last dimension of the input. Given this condition, it was easy to implement, so I created it.

good luck with your work.

```php
<?php
namespace Rindow\NeuralNetworks\Builder;
class Gradient
{
    public function softmax(
        NDArray $x,
    ) : NDArray
    {
        $func = new Softmax($this->backend);
        return $func($x);
    }

}
```


```php
<?php
namespace Rindow\NeuralNetworks\Gradient\Func;

use InvalidArgumentException;
use Interop\Polite\Math\Matrix\NDArray;
use Rindow\NeuralNetworks\Gradient\Core\AbstractFunction;

class Softmax extends AbstractFunction
{
    protected function call(array $inputs) : array
    {
        $K = $this->backend;
        $container = $this->container();
        $outputs = $K->softmax($inputs[0]);
        $container->outputs = $outputs;
        return [$outputs];
    }

    protected function differentiate(array $dOutputs) : array
    {
        $K = $this->backend;
        $container = $this->container();
        $dInput = $K->dSoftmax($dOutputs[0], $container->outputs);
        return [$dInput];
    }
}
```

```php
<?php
namespace RindowTest\NeuralNetworks\Gradient\Func\SoftmaxTest;

use PHPUnit\Framework\TestCase;
use Rindow\Math\Matrix\MatrixOperator;
use Rindow\NeuralNetworks\Builder\NeuralNetworks;
use Interop\Polite\Math\Matrix\NDArray;

class SoftmaxTest extends TestCase
{
    public function newMatrixOperator()
    {
        return new MatrixOperator();
    }

    public function newNeuralNetworks($mo)
    {
        return new NeuralNetworks($mo);
    }

    public function newBackend($nn)
    {
        return $nn->backend();
    }

    public function testLossOne()
    {
        $mo = $this->newMatrixOperator();
        $nn = $this->newNeuralNetworks($mo);
        $K = $this->newBackend($nn);
        $g = $nn->gradient();

        $x = $g->Variable($K->array([
            [1.0, 2.0, 3.0],
            [0.5, 1.5, 2.5]
        ],dtype:NDArray::float32));

        $y = $nn->with($tape=$g->GradientTape(),
            function() use ($g,$x){
                $y = $g->softmax($x); // always axis=-1
                return $y;
            }
        );
        $dx = $tape->gradient($y,$x);

        $y = $K->ndarray($y->value());
        $truesY = $mo->array([
            [0.090031, 0.244728, 0.665241],
            [0.090031, 0.244728, 0.665241],
        ]);
        $this->assertTrue($mo->la()->isclose(
            $y,
            $truesY,
        ));

        $truesDx = $mo->array([
            [0, 0, 0],
            [0, 0, 0],
        ]);
        $this->assertTrue($mo->la()->isclose(
            $dx,
            $truesDx,
        ));
    }

    public function testNormal()
    {
        $mo = $this->newMatrixOperator();
        $nn = $this->newNeuralNetworks($mo);
        $K = $this->newBackend($nn);
        $g = $nn->gradient();

        $x = $g->Variable($K->array([
            [1.0, 2.0, 3.0],
            [0.5, 1.5, 2.5]
        ],dtype:NDArray::float32));
        $salt = $mo->la()->range(start:1,limit:1+array_product($x->shape()),dtype:NDArray::float32)
                ->reshape($x->shape());
        $salt = $g->Variable($salt);

        [$outputs,$y] = $nn->with($tape=$g->GradientTape(),
            function() use ($g,$x,$salt){
                $y = $g->softmax($x); // always axis=-1
                $outputs = $g->mul($y,$salt);
                return [$outputs,$y];
            }
        );
        $dx = $tape->gradient($outputs,$x);

        $y = $K->ndarray($y->value());
        $truesY = $mo->array([
            [0.090031, 0.244728, 0.665241],
            [0.090031, 0.244728, 0.665241],
        ]);
        $this->assertTrue($mo->la()->isclose(
            $y,
            $truesY,
        ));

        $truesDx = $mo->array([
            [-0.1418171 , -0.14077035,  0.28258747],
            [-0.14181706, -0.14077029,  0.28258765],
        ]);
        $this->assertTrue($mo->la()->isclose(
            $dx,
            $truesDx,
        ));
    }
}
```

**Comment by zhuruicheng** [comment]
Thanks for your reply, I can use it normally now!
